cddl/any-of.cddl

```diff
@@ -0,0 +1 @@
+any-of<T> = T / #6.562([ + T])
```

Comment:
It may be worth trying to frame this more generally in terms of "matchers", of which `any-of` would be an instance.

For example, we could make the type of each claim value a `$type-matcher`:

```
measurement-values-map = non-empty<{
  ? svn: $svn-matchers
  ? digest: $digest-matchers
  ; ...
}>
```

A matcher would be defined as a choice between different high-level matcher types, e.g., the _naked_ type, a _multiple_ type (similar to `any-of` here), a _range_ type, etc., depending on what makes sense for that claim.

```
match-multiple<T> = #6.554([ 2* T ])
match-range<T> = #6.555(range-desc<T>)
; match-regex, match-masked-bytes, etc.
```


For example, for SVNs, it makes sense to define all three above:

```
$svn-type = uint

$svn-matchers /= $svn-type
$svn-matchers /= match-multiple<$svn-type>
$svn-matchers /= match-range<$svn-type>
```

For digests (typically, random binary blobs) it makes little sense to add anything more than the "multiple" matcher:

```
$digest-type = [ val: bytes, alg: int ]

$digest-matchers /= $digest-type
$digest-matchers /= match-multiple<$digest-type>
```

etc.

Reply:
"matcher" is more often called a "pattern" in programming languages (as in "pattern matching"). I could definitely be down for framing the values as having their own generic or specific pattern forms, but I don't thing making them extensible is necessarily the right choice. If there are profile-specific representation choices, then they should use the values map extension socket, IMHO.

Reply:
In this regard, raw-value-mask could be a specific pattern for raw-value, such that a bstr matches a pattern #6.TBD(mask, raw-value-after-zeroextended-mask) instead of needing two values, but that representation choice is only good if the mask isn't needed on its own for lookup.

Reply:
> "matcher" is more often called a "pattern" in programming languages (as in "pattern matching").

WFM, **Pattern Matching** looks like a good section title :-D

> I could definitely be down for framing the values as having their own generic or specific pattern forms, but I don't think making them extensible is necessarily the right choice.

The proposal is not to make them extensible.
The proposal is to define a few useful pattern-matching constructs in terms of their semantics, rules and CBOR tagging.  Then define the types of a claim in terms of a _fixed_ type choice that is built from those blocks.

> If there are profile-specific representation choices, then they should use the values map extension socket, IMHO.

Yes.

Reply:
If I characterize this vs. what is described in [Intel Profile](https://github.com/nedmsmith/draft-cds-rats-intel-corim-profile/tree/main). This uses CDDL macros to define "possible state" while Intel Profile uses CDDL directly. Use of CDDL directly is possibly more flexible as it is not limited by assumptions of the macro processor. For example, any-of<> models subset semantics, but what about superset and disjoint set? These are use cases that Intel would like to support. 

The Intel Profile supports numeric ranges (greater-than, less-than, greater-equal, less-equal, inequality etc). The above suggests there could be a range macro,  but didn't define it, so it isn't clear that it would have the same utility as the CDDL definition. 

The Intel Profile also supports time and epoch types. 

I think the use of macros is interesting and creative, but why is it better than using CDDL directly?

Reply:
any-of is not subset semantics, I think it's your <evidence_object><member><reference_set>, where the reference set is the [ + T ], and evidence_object is provided by the ECT, if I understand correctly.

If this is the case, then I'd want to allow for these set expressions in any reference value. Where you have set-type = [ * any ], I'd suggest set-type<T> = [ * T ].